fv/latency_test.go

```diff
@@ -0,0 +1,230 @@
+		felix    *containers.Felix
+		felixPID int
+		client   client.Interface
+
+		resultsFile *os.File
+	)
+
+	BeforeEach(func() {
+		topologyOptions := containers.DefaultTopologyOptions()
+		felix, etcd, client = containers.StartSingleNodeEtcdTopology(topologyOptions)
+		felixPID = felix.GetFelixPID()
+
+		// Install the hping tool, which we use for latency measurments.
+		felix.Exec("sh", "-c", "echo http://dl-cdn.alpinelinux.org/alpine/edge/testing >> /etc/apk/repositories")
+		felix.Exec("apk", "update")
+		felix.Exec("apk", "add", "hping3")
+
+		var err error
+		resultsFile, err = os.OpenFile("latency.log", os.O_WRONLY|os.O_APPEND|os.O_CREATE, 0644)
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	AfterEach(func() {
+		err := resultsFile.Close()
+		if err != nil {
+			log.WithError(err).Error("Close returned error")
+		}
+
+		if CurrentGinkgoTestDescription().Failed {
+			felix.Exec("iptables-save", "-c")
+		}
+		felix.Stop()
+
+		if CurrentGinkgoTestDescription().Failed {
+			etcd.Exec("etcdctl", "ls", "--recursive", "/")
+		}
+		etcd.Stop()
+	})
+
+	describeLatencyTests := func(c latencyConfig) {
+		var (
+			w   [2]*workload.Workload
+			cc  *workload.ConnectivityChecker
+			pol *api.GlobalNetworkPolicy
+		)
+
+		createPolicy := func(policy *api.GlobalNetworkPolicy) *api.GlobalNetworkPolicy {
+			log.WithField("policy", dumpResource(policy)).Info("Creating policy")
+			policy, err := client.GlobalNetworkPolicies().Create(utils.Ctx, policy, utils.NoOptions)
+			Expect(err).NotTo(HaveOccurred())
+			return policy
+		}
+
+		updatePolicy := func(policy *api.GlobalNetworkPolicy) *api.GlobalNetworkPolicy {
+			log.WithField("policy", dumpResource(policy)).Info("Updating policy")
+			policy, err := client.GlobalNetworkPolicies().Update(utils.Ctx, policy, utils.NoOptions)
+			Expect(err).NotTo(HaveOccurred())
+			return policy
+		}
+
+		BeforeEach(func() {
+			for ii := range w {
+				iiStr := strconv.Itoa(ii)
+				var ports string
+
+				ports = "3000"
+				w[ii] = workload.Run(
+					felix,
+					"w"+iiStr,
+					"cali0"+iiStr,
+					c.workloadIP(ii),
+					ports,
+					"tcp",
+				)
+
+				w[ii].DefaultPort = "3000"
+				w[ii].Configure(client)
+			}
+
+			cc = &workload.ConnectivityChecker{
+				Protocol: "tcp",
+			}
+
+			pol = api.NewGlobalNetworkPolicy()
+			pol.Namespace = "fv"
+			pol.Name = "policy-1"
+			pol.Spec.Ingress = []api.Rule{
+				{
+					Action: "Allow",
+				},
+			}
+			pol.Spec.Egress = []api.Rule{
+				{
+					Action: "Allow",
+				},
+			}
+			pol.Spec.Selector = "all()"
+
+			pol = createPolicy(pol)
+
+			cc.ExpectSome(w[0], w[1])
+			cc.ExpectSome(w[1], w[0])
+			cc.CheckConnectivity()
+		})
+
+		It("with allow-all should have good latency", func() {
+			meanRtt := w[0].LatencyTo(w[1].IP, w[1].DefaultPort)
+			_, err := fmt.Fprintf(resultsFile, "allow-all: %v\n", meanRtt)
+			Expect(meanRtt).To(BeNumerically("<", 10*time.Millisecond))
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		Describe("with all() selector", func() {
```

Comment:
Saying "source selector" would be a touch clearer, given that there's already another use of `all()` above.